oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+unique content to be shown.
+
+Prototype: Discussions
+----------------------
+
+The prototype switched the "Discussion" tab to use the `ComponentTabMixin`,
+and then provided a new `DiscussionBoardComponentView` class. Providing this
+tab as a component allows the content to be rendered as AJAX or within an
+iframe or mobile web view. In particular, this could allow new LMS front end
+implementations to be provided that don't render with the same edX-style tab
+interface. Such a front end could now include discussions anywhere that it
+chooses.
+
+Prototype: Instructor Dashboard
+-------------------------------
+
+The final piece of the prototype demonstrates a dynamically generated
+instructor dashboard built purely using component views. The dashboard home
+page makes a request to find all installed instructor component and then
+displays them as a set of cards organized by category:
+
+.. image:: oep-0012/proposed-instructor-dashboard.png
+  :alt: A screen shot from the instructor dashboard prototype, showing
+  how each feature that can be controlled is shown as a card that can
+  be drilled down into to be seen full screen.
+
+Each card has a "View" link that lets the user drill down and see the
+component for the feature being represented. This component is shown full
+screen but with a back button to allow the user to navigate back to the
+dashboard home.
+
+Rationale
+=========
+
+Why are we not using an industry framework?
+-------------------------------------------
+
+It seems that the requirements for pluggable user interfaces are not unique
+to Open edX, and hence other solutions are likely to exist. Our research so
+far has not found a solution that covers the full set of use cases, but that
+might change in the future.
+
+One approach that is similar is Pinterest's solution that can partially render a
+page, and then asynchronously request the missing components and fill them in as
+they become available. They use Mozilla's Nunjucks templating library in Django
+and in particular their `Nunjucks Asynchronous Support`_. However, it appears
+that this only supports the ability to asynchronously return HTML, and doesn't
+allow for the ability to dynamically add CSS and JavaScript. See their blog
+post `How we switched our template rendering engine to React`_ for more
+details.
+
+Another similar use case is HTML5 ads, but those are all implemented as
+iframes. Using iframes is possible for some pluggable situations (such as LTI),
+but it would provide a poor user experience if a complex page was constructed
+from tens or even hundreds of iframes.
+
+Could we use XBlocks instead of introducing Django Component Views?
+-------------------------------------------------------------------
+
+Note that all of the requirements of pluggable user interfaces are supported by
+XBlocks. However, there are several reasons why XBlocks are not always
+appropriate:
+
+* XBlocks need to have a unique usage id so that instances can be referenced.
```

Comment:
So do the plug-in components, actually (see above discussion about unique IDs) :)